File: Medium/213. House Robber II.py
class Solution:
    def rob(self, nums: list) -> int:
        if not nums: return 0
        if len(nums) <= 2: return max(nums)
        def notCircle(nums):
            pre,cur = 0,0
            
            for num in nums:
                tmp = cur
                cur = max(cur, num+pre)
                pre = tmp
                
            return cur
        
        left = notCircle(nums[:len(nums)-1])
        right = notCircle(nums[1:]) 
        
        return max(left,right)

File: Medium/213. House Robber II.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 0) return 0;
		
		if(nums.size() <  2){
			return *std::max_element(nums.begin(), nums.end());
		}
		
		vector<int> left(&nums[0],&nums[nums.size()-1]);
		vector<int> right(&nums[1],&nums[nums.size()-0]);
		
		return std::max(notCircle(left), notCircle(right));
		
    }
	
	int notCircle(vector<int>& nums){
		int pre = 0, cur = 0;
		
		for(int i = 0; i < nums.size(); i++){
			int tmp = cur;
			cur = std::max(cur, nums[i] + pre);
			pre = tmp;
		}
		return cur;
	}
};